ProcessThreadTests.TestStartTimeProperty - Get spawned thread by ID from Process.Threads to avoid mismatch

adamsitnik opened the pull request:
fixes #103448

adamsitnik committed:
Increasing the tolerance for Mono Runtime for the flaky ProcessThreadTests.TestStartTimeProperty test, fixes #103448

danmoseley commented:
Although, I think @Jozkee has a point in his comment in the issue, that tolerating such wrong timestamps is probably masking a real bug rather than just accommodating test noise.

adamsitnik committed:
select the thread based on current thread native id, remove the try catch block:

- it can hide LINQ issues as LINQ may throw InvalidOperationException
- it's not needed for the current thread, because the code is still running so we know the thread is alive

adamsitnik committed:
Revert "Increasing the tolerance for Mono Runtime for the flaky ProcessThreadTests.TestStartTimeProperty test, fixes #103448"

This reverts commit 3092d861540bcbc3c4fe341f160a5a22bc02c102.

danmoseley commented:
Hmm, is the likely reason for the failure that `Task.Factory.StartNew(() =>` cannot be assumed to start a new thread, even with TaskCreationOptions.LongRunning ? If that's the case, I don't think your change will fix that problem. The fix would be to actually create a new thread (?)

Jozkee commented on src/libraries/System.Diagnostics.Process/tests/ProcessThreadTests.cs:
```diff
@@ -148,17 +149,22 @@ public async Task TestStartTimeProperty()
             }
+
+            [DllImport("libc")]
+            static extern int gettid();
```
[CI failures](https://[helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-104972-merge-91f33105ce7d418cab/System.Diagnostics.Process.Tests/1/console.e209f89d.log?helixlogtype=result](https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-104972-merge-91f33105ce7d418cab/System.Diagnostics.Process.Tests/1/console.e209f89d.log?helixlogtype=result)) are related to this method. `System.EntryPointNotFoundException : gettid`.

Jozkee commented on src/libraries/System.Diagnostics.Process/tests/ProcessThreadTests.cs:
```diff
@@ -148,17 +149,22 @@ public async Task TestStartTimeProperty()
+
+            [DllImport("kernel32.dll")]
+            [SuppressGCTransition]
+            static extern int GetCurrentThreadId();
```
We already have this method for Windows? How was this able to compile?
https://github.com/dotnet/runtime/blob/3f607bd6061299a20d29b76914b4176b9d835f66/src/libraries/System.Diagnostics.Process/tests/ProcessThreadTests.Windows.cs#L55-L56

Jozkee commented on src/libraries/System.Diagnostics.Process/tests/ProcessThreadTests.cs:
```diff
@@ -148,17 +149,22 @@ public async Task TestStartTimeProperty()
                     p.Refresh();
-                    try
-                    {
-                        var newest = p.Threads.Cast<ProcessThread>().OrderBy(t => t.StartTime.ToUniversalTime()).Last();
```
Was it wrongly assumed that Process.Threads contained the threads ordered with the most recently created as the `Last()` one? 

Is also possible that the property does have the threads ordered but StartNew wasn't creating a new thread but reusing an existing one as Dan mentioned.

adamsitnik replied:
> Was it wrongly assumed that Process.Threads contained the threads ordered with the most recently created as the Last() one?

There are other tests running in parallel and each of them can create a new thread. Even if we disable that, the `ThreadPool` can still start a new thread on it's own. So to get 100% deterministic behavior we should IMO perform an ID match.

adamsitnik replied:
> StartNew wasn't creating a new thread but reusing an existing one as Dan mentioned.

In theory CLR and Mono are sharing the same thread pool implementation, so I would expect `TaskCreationOptions.LongRunning` to always create a new thread, but I might be wrong.

@kouvel is it safe to assume that `Task.Factory.StartNew(/**/, CancellationToken.None, TaskCreationOptions.LongRunning, TaskScheduler.Default);` executed from an xUnit test runner will always spawn a new thread?

kouvel replied:
Based on this:
https://github.com/dotnet/runtime/blob/27776e2fbcf865029829027bef9a272326e70105/src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/ThreadPoolTaskScheduler.cs#L45-L53

It looks like when thread start is supported, a new thread is created for each long-running task. The threads used for long-running tasks are not pooled currently, though they could be. Probably creating a new thread would be more explicit in intent.

Wonder if the failure is happening when thread start is not supported. In the threading tests that create new threads, the following condition is used, which appears to be the same condition as `Thread.IsThreadStartSupported`:
`[ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsThreadingSupported))]`

kouvel replied:
Although using a long-running task would avoid an unhandled exception/crash on assertion failure in the new thread. `ThreadTestHelpers.CreateGuardedThread` could be used for, as in: https://github.com/dotnet/runtime/blob/27776e2fbcf865029829027bef9a272326e70105/src/libraries/System.Threading.Thread/tests/ThreadTests.cs#L621

adamsitnik replied:
@kouvel thank you!

danmoseley replied:
> The threads used for long-running tasks are not pooled currently, though they could be.

TIL - I should have checked the code. @kouvel is that worth an issue, as a suggestion? Basically pull a thread from the pool if available, and trigger a replacement to start. I have no idea.

kouvel replied:
What's done currently is simple and may be sufficient if most long-running work items rarely or never complete. If there's a need where work items would complete frequently enough, and are not short enough and need to be classified as long-running, then it may be worthwhile to pool those threads.

adamsitnik committed:
fix the Linux build: gettid is not exposed in older versions of libc

adamsitnik committed:
Make the test conditional

jozkee reviewed (approved):
LGTM, assuming CI passes, thanks!

adamsitnik commented on src/libraries/System.Diagnostics.Process/tests/ProcessThreadTests.cs:
```diff
@@ -148,15 +149,13 @@ public async Task TestStartTimeProperty()
+                    ProcessThread[] processThreads = p.Threads.Cast<ProcessThread>().ToArray();
+                    ProcessThread newThread = Assert.Single(processThreads, thread => thread.Id == newThreadId);
+
+                    Assert.InRange(newThread.StartTime.ToUniversalTime(), curTime - allowedWindow, DateTime.Now.ToUniversalTime() + allowedWindow);
```
I've reverted the increase of the tolerance. I think it's safer to keep the current one (I am just afraid it would make the test even more flaky)

Jozkee replied:
I think with your fix the test won't be flaky at all 😆 but I might be wrong.

adamsitnik committed:
get SYS_gettid value specific to given architecture (and hopefully fix the build for arm64)
